internal/pkg/scaffold/cmd.go

```diff
@@ -146,9 +151,11 @@ func main() {
 	}
 
 	// Create Service object to expose the metrics port.
-	_, err = metrics.ExposeMetricsPort(ctx, metricsPort)
-	if err != nil {
-		log.Info(err.Error())
+	if !*disableMetrics {
```

Comment:
hmm why is this change needed?

Reply:
This is needed because the leader election and metrics features check `/var/run/secrets/kubernetes.io/serviceaccount/namespace` to determine whether or not they are running outside of a cluster and should be disabled. This works in a large majority of cases, but it doesn't work in Openshift-CI because the tests are run from a pod in an Openshift cluster. This causes metrics and leader election to try and start during the `up local` test, which breaks the test. I added the `disable-metrics` and `disable-leader-election` flags so we could manually disable these features during the `up local` tests: https://github.com/operator-framework/operator-sdk/pull/1384/files#diff-62b541caac15c49bbed3723129bf478eR510

Reply:
I understand now, but is it the right thing to do, to add those just for tests to pass?

Reply:
Can't we just modify the files in the tests instead? I would prefer that compared to adding these flags to all the users, we signal to them these are things they should be disabling. WDYT?

Reply:
I agree with Lili. I don't think we should make these changes in the scaffold. I'd second the idea of investigating whether we can change or move the `/var/run/secrets/kubernetes.io/` files during the test.

Reply:
We cannot change `/var/run/secrets/kubernetes.io/serviceaccount` as that's mounted read-only.

Reply:
Ooof. Yeah, forgot about that. What about overriding the kubeconfig that is used? It looks like the in-cluster config has low precedence and could be overridden with an environment variable or flag.

https://github.com/kubernetes-sigs/controller-runtime/blob/b88347ba72e1ba07ed3bed2fcca1ac0633f76875/pkg/client/config/config.go#L47-L62

Reply:
The issue isn't the kubeconfig; that points to the external cluster that the operator is watching and we need that for the tests. The problem is that the tests themselves are running in the openshift-CI cluster and thus have the serviceaccount path as well which makes the local test believe it's running in the cluster instead of locally.

Reply:
Technically, the part that gives us problems is looking up the `POD_NAME` env var, as that doesn't exist in this pod. That check occurs here: https://github.com/operator-framework/operator-sdk/blob/master/pkg/k8sutil/k8sutil.go#L95:L97

One way we could solve this would be to have a specific env var we could set that causes the check to exit with a different error type that the metrics and leader election functions can use to determine that they shouldn't start. We could use something that wouldn't actually match a pod name regex to make sure a user doesn't accidentally trigger it.

Reply:
More generally, is the issue that we're trying to run a local test within a pod running in a cluster, and we make assumptions about the environment (or lack thereof) for running locally that don't hold in this scenario?

By the time we've gotten around to calling `GetPod`, we've already decided (incorrectly) that we're running in `cluster` mode, not `local` mode. So the question in my mind is how can we hint to the process what mode it's in? Right now that's handled somewhat magically but looking things up in the filesystem and environment. Maybe we need to be more explicit?

Reply:
One way we could be explicit would be to add a new flag/arg that `up local` (and `test local --up-local`) always sets to tell the operator that it is running in local mode. That solution would be similar to what this PR has, but it would be a single flag to disable in-cluster features like leader election, metrics, and any other cluster only features we may add in the future.

Reply:
That would require the flag to be somewhat prominently added to the binary, which I think is what Lili is suggesting we avoid.

Maybe our `k8sutil` functions that care about where the process is running can check an environment variable (e.g. `OSDK_FORCE_RUN_MODE=(local|cluster)`), and fallback to the existing detection logic if it's not set. I realize that's still a little bit "magic" and implicit, but maybe that strikes a better balance?

Maybe we could even do something like the following to consolidate all of this detection logic?

```go
package k8sutil

type RunMode string

const (
    RunModeLocal RunMode = "local"
    RunModeCluster RunMode = "cluster"
)

type Environment struct {
    Mode RunMode
    Namespace string
    Pod *corev1.Pod
}

func GetEnvironment() (e *Environment, err error) {
    e = &Environment{}
    if forceMode, ok := os.LookupEnv(RunModeEnv); ok {
        e.RunMode = RunMode(forceMode)
        if e.RunMode != RunModeLocal && e.RunMode != RunModeCluster {
            return nil, fmt.Errorf("invalid run mode %s", forceMode)
        }
    } else {
        // autodetect mode from files/env
    }
    if e.RunMode == "cluster" {
        e.Pod, err = GetPod()
        handleErr(err)
        e.Namespace, err = GetNamespace()
        handleErr(err)
    }
    return e, nil
}
```

Reply:
Ehh. Maybe that's overkill 🤷‍♂. I was thinking of being able to do that in an init function and setting a global variable in the `k8sutil` package for other functions to use, to avoid repeatedly doing the same operations. But consolidating maybe doesn't buy a whole lot. 

Anyways, WDYT of using just an environment variable and doing the checks within `k8sutil`?

Reply:
For now I think setting an `OSDK_FORCE_RUN_MODE` env var and just handling that in the current functions would be good. Maybe in another PR we could do a bit of refactoring by adding a `GetEnvironment` function to simplify some of the code.

Reply:
I've updated the PR to use the env vars instead and have successfully tested in openshift-ci